src/Microsoft.AspNetCore.JsonPatch/Internal/DynamicObjectAdapter.cs

```diff
@@ -0,0 +1,221 @@
+            if (property.GetType().GetTypeInfo().IsValueType
+                && Nullable.GetUnderlyingType(property.GetType()) == null)
+            {
+                value = Activator.CreateInstance(property.GetType());
+            }
+
+            if (!TrySetDynamicObjectProperty(target, contractResolver, segment, value))
+            {
+                errorMessage = Resources.FormatTargetLocationAtPathSegmentNotFound(segment);
+                return false;
+            }
+
+            errorMessage = null;
+            return true;
+
+        }
+
+        public bool TryReplace(
+            object target,
+            string segment,
+            IContractResolver contractResolver,
+            object value,
+            out string errorMessage)
+        {
+            object property = null;
+            if (!TryGetDynamicObjectProperty(target, contractResolver, segment, out property))
+            {
+                errorMessage = Resources.FormatTargetLocationAtPathSegmentNotFound(segment);
+                return false;
+            }
+
+            object convertedValue = null;
+            if (!TryConvertValue(value, property.GetType(), out convertedValue))
+            {
+                errorMessage = Resources.FormatInvalidValueForProperty(value);
+                return false;
+            }
+
+            if (!TrySetDynamicObjectProperty(target, contractResolver, segment, convertedValue))
+            {
+                errorMessage = Resources.FormatTargetLocationAtPathSegmentNotFound(segment);
+                return false;
+            }
+
+            errorMessage = null;
+            return true;
+        }
+
+        public bool TryTraverse(
+            object target,
+            string segment,
+            IContractResolver contractResolver,
+            out object nextTarget,
+            out string errorMessage)
+        {
+            var dynamicObject = target as IDynamicMetaObjectProvider;
+            if (dynamicObject == null)
+            {
+                nextTarget = null;
+                errorMessage = null;
+                return false;
+            }
+
+            object property = null;
+            if (!TryGetDynamicObjectProperty(target, contractResolver, segment, out property))
+            {
+                nextTarget = null;
+                errorMessage = null;
+                return false;
+            }
+            else
+            {
+                nextTarget = property;
+                errorMessage = null;
+                return true;
+            }
+        }
+
+        private bool TryGetDynamicObjectProperty(
+            object target,
+            IContractResolver contractResolver,
+            string segment,
+            out object property)
+        {
+            var jsonDynamicContract = contractResolver.ResolveContract(target.GetType()) as JsonDynamicContract;
+
+            var propertyName = jsonDynamicContract?.PropertyNameResolver(segment);
```

Comment:
You'll get an ArgumentNullException later at line 156. If you are going to be defensive, make sure you are defending from the right things